### PR TITLE
chore: migrate more files to TS

### PR DESCRIPTION
**Pull request by zamotany**
* `src/index.js` was not migrated as it's breaking the tests, probably something with resolving linaria import is working as expected
* `src/react/*.js` - same story, even changing the extension causes tests to fail

@satya164 do you have any idea why that is happening? Or how we can solve that?

### CODE REVIEW

**Comment by satya164** [comment]
I'll check tonight or in Tuesday

File: babel.js [hunk 1]
```diff
@@ -1,3 +1,3 @@
 /* eslint-disable import/no-unresolved */
 
-module.exports = require('./lib/babel');
+module.exports = require('./lib/babel').default;
```
Comment (Anber) [inline babel.js]: This line broke my canary build.

Comment (zamotany) [reply babel.js]: Yeah I forgot to refactor this line to use ES modules: https://github.com/callstack/linaria/blob/chore/migrate-to-fs/src/babel/index.ts#L3

**Comment by pbitkowski** [comment]
@zamotany could you include changes from #377 if they make sense?

**Comment by thymikee** [comment]
FYI: https://github.com/callstack/linaria/pull/387

**Comment by zamotany** [comment]
@satya164 @pbitkowski @thymikee 
Finished migration, ported those 2 PRs, everything is passing, website is working just fine, but:
* testing declaration files is not yet done
* stylelint is working - it's not picking up any errors

I probably won't have time to deal with that in the next days, so if someone wants to do it, feel free.

File: src/loader.ts [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
+  this: any,
+  content: string,
+  inputSourceMap: Object | null
+) {
   const {
-    sourceMap,
+    sourceMap = undefined,
     cacheDirectory = '.linaria-cache',
-    preprocessor,
+    preprocessor = undefined,
```
Comment (thymikee) [inline src/loader.ts]: this assignment will only happen once `processor` (or `sourceMap` above) is undefined, so this logic is redundant

File: src/react/styled.ts [hunk 3]
```diff
@@ -1,21 +1,19 @@
...
+  name: string;
+  class: string;
   vars?: {
-    [string]: [
-      string | number | ((props: *) => string | number),
-      string | void,
-    ],
-  },
+    [key: string]: [
+      string | number | ((props: any) => string | number),
```
Comment (thymikee) [inline src/react/styled.ts]: prefer unknown?
```suggestion
      string | number | ((props: unknown) => string | number),
```

File: src/react/styled.ts [hunk 4]
```diff
@@ -46,14 +44,14 @@ function styled(tag: React.ComponentType<*> | string) {
       }
     }
 
-    const render = (props, ref) => {
+    const render = (props: any, ref: any) => {
```
Comment (thymikee) [inline src/react/styled.ts]: something like this maybe?
```suggestion
    const render = (props: { [key: string]: unknown }, ref: React.RefObject<unknown>) => {
```
not sure if ref object is different for function refs and createRefs

File: src/react/styled.ts [hunk 5]
```diff
@@ -121,32 +117,42 @@ function styled(tag: React.ComponentType<*> | string) {
...
-|};
+export default ((process.env.NODE_ENV !== 'production'
+  ? new Proxy(styled, {
+      get(o, prop) {
+        prop = typeof prop === 'number' ? prop.toString() : prop;
+        prop = typeof prop === 'symbol' ? prop.toString() : prop;
+        return o(prop);
+      },
+    })
+  : styled) as any) as Styled;
```
Comment (thymikee) [inline src/react/styled.ts]: can't we just cast the Proxy and mark `styled as Styled`?

File: src/server/collect.ts [hunk 6]
```diff
@@ -45,7 +43,7 @@ const collect = (html: string, css: string): CollectResult => {
   };
 
   stylesheet.walkRules(rule => {
-    if (rule.parent.name === 'keyframes') {
+    if ('name' in rule.parent && rule.parent.name === 'keyframes') {
```
Comment (thymikee) [inline src/server/collect.ts]: nice, these are likely bugfixes

File: src/transform.ts [hunk 7]
```diff
@@ -80,22 +80,25 @@ module.exports = function transform(code: string, options: Options): Result {
       sourceFileName: options.filename,
       inputSourceMap: options.inputSourceMap,
     }
-  );
+  )!;
 
-  if (!metadata.linaria) {
+  if (!metadata || !(metadata as any).linaria) {
```
Comment (thymikee) [inline src/transform.ts]: I think we could extend the metadata type returned by babel and avoid casting here and there

File: src/transform.ts [hunk 8]
```diff
@@ -117,7 +121,7 @@ module.exports = function transform(code: string, options: Options): Result {
                 // Replace asset path with new path relative to the output CSS
                 path.relative(
                   /* $FlowFixMe */
```
Comment (thymikee) [inline src/transform.ts]: remove $FlowFixMe (here and everywhere)

File: src/loader.ts (hunk 2 above)
Comment (zamotany) [reply src/loader.ts]: nope, TS will cry that it has no default initialiser

Comment (thymikee) [reply src/loader.ts]: srsly, that's dumb

File: src/react/styled.ts (hunk 4 above)
Comment (zamotany) [reply src/react/styled.ts]: I have no idea what the correct types are for this file

Comment (thymikee) [reply src/react/styled.ts]: let's leave it for later then, no worries. marking a `// TODO: better types` would do just fine :)

File: src/react/styled.ts (hunk 5 above)
Comment (zamotany) [reply src/react/styled.ts]: `styled` also needs to be casted to `any` and then to `styled`. Ideally it should not be needed and the types above should be integrated into implementation, but it's time consuming

File: src/server/collect.ts (hunk 6 above)
Comment (zamotany) [reply src/server/collect.ts]: technically `'name' in rule.parent ` is here to only make TS happy XD

File: src/transform.ts (hunk 8 above)
Comment (zamotany) [inline src/transform.ts]:
```suggestion
```

File: src/server/collect.ts (hunk 6 above)
Comment (thymikee) [reply src/server/collect.ts]: I remember something about some of the proprties being undefined while we assumed they don't. Found it: https://github.com/callstack/linaria/issues/217. Likely not related, but still better to be safe

File: src/react/styled.ts (hunk 5 above)
Comment (thymikee) [reply src/react/styled.ts]: yea that's what I'm concerned about. feel free to mark as todo for later

File: src/react/styled.ts (hunk 4 above)
Comment (zamotany) [reply src/react/styled.ts]: Added a `// TODO` comment for whole `styled` component

**Comment by Anber** [comment]
Let's merge it all and release first alpha! :)

**Pull request merged**